Studies/source/PuritySignalStudy.cxx

```diff
@@ -0,0 +1,240 @@
+            histos.at(histo_index).SetLineWidth(2);
+            histos.at(histo_index).Draw("same");
+
+            legend->AddEntry(&histos.at(histo_index), ToString(histo_names.at(histo_index)).c_str());
+            legend->SetTextSize(0.022f);
+            legend->Draw("same");
+        }
+        gPad->SetLeftMargin(0.15f);
+        gPad->Update();
+
+        canvas.Print(("match_plots_" + ToString(args.channel()) + ".pdf").c_str(), "Title:jet_ordering");
+    }
+
+    size_t calculateMatchesJets(const Event& event, JetOrdering jet_ordering)
+    {
+        auto ordered_jets = orderJets(event, jet_ordering);
+
+        const size_t n_max = std::min(static_cast<int>(ordered_jets.size()), 2);
+        size_t count = 0;
+        for(size_t n = 0; n < n_max; ++n) {
+            const size_t index = ordered_jets.at(n).index;
+            if(event.jets_genJetIndex.at(index) != -1) ++count;
+        }
+        anaData.jet_matches(jet_ordering).Fill(count);
+        return count;
+    }
+
+    size_t calculateMatchesTaus(const Event& event, TauIdDiscriminator tau_id_discriminator, bool is_descending_order = true)
+    {
+        std::vector<size_t> ordered_indexes = orderTaus(event, tau_id_discriminator, is_descending_order);
+        //Cases for eTau & muTau with zero taus remaning after applied cut
+
+        const size_t n_expected = args.channel() == Channel::TauTau ? 2 : 1;
+        const size_t n_max = std::min(ordered_indexes.size(), n_expected);
+        size_t count = 0;
+        for(size_t n = 0; n < n_max; ++n) {
+            const size_t index = ordered_indexes.at(n);
+            if(event.lep_genTauIndex.at(index) != -1) ++count;
+        }
+        anaData.lep_matches(tau_id_discriminator).Fill(count);
+        return count;
+    }
+
+private:
+
+    std::vector<analysis::jet_ordering::JetInfo<LorentzVector>> orderJets(const Event& event, JetOrdering jet_ordering)
+    {
+        std::vector<analysis::jet_ordering::JetInfo<LorentzVector>> jet_info_vector;
+        BTagger bTagger(Parse<analysis::Period>(args.period()) ,jet_ordering);
+
+        for(size_t jet_index = 0; jet_index < event.jets_p4.size(); ++jet_index)
+            jet_info_vector.emplace_back(LorentzVectorE(event.jets_p4.at(jet_index)), jet_index, bTagger.BTag(event,jet_index));
+
+        return jet_ordering::OrderJets(jet_info_vector, true, cuts::btag_2017::pt, cuts::btag_2017::eta);
+    }
+
+    std::vector<size_t> orderTaus(const Event& event, TauIdDiscriminator tau_id_discriminator, bool is_descending_order = true)
+    {
+        std::vector<float> raw_values;
+        for(size_t lep_index = 0; lep_index < event.lep_p4.size(); ++lep_index){
+           if (static_cast<LegType>(event.lep_type.at(lep_index)) == analysis::LegType::tau ) {
+               ntuple::TupleLepton lepton(event, lep_index);
+
+               const bool is_good_lepton = (args.channel() == Channel::ETau && lepton.Passed(TauIdDiscriminator::byDeepTau2017v2VSe,DiscriminatorWP::VVVLoose)) ||
```

Comment:
In order to exclude the part of the parameter space where the level of mis-modeling can be quite high, it is better if VVVLoose against electron and VLoose against muon are applied for all the channels.